src/vc_util/src/lib.rs

```diff
@@ -350,52 +384,65 @@ fn validate_expiration(
+//     cf. a convention at https://github.com/dfinity/internet-identity/blob/main/docs/vc-spec.md#recommended-convention-connecting-credential-specification-with-the-returned-credentials
 fn validate_claims_match_spec(
     vc_claims: &Map<String, Value>,
     spec: &CredentialSpec,
 ) -> Result<(), JwtValidationError> {
-    let credential_type = vc_claims
+    let credential_type = &spec.credential_type;
+
+    // Check that type-claim contains spec.credential_type.
+    let vc_type_entry = vc_claims
         .get("type")
-        .ok_or(inconsistent_jwt_claims("missing type JWT vc"))?;
-    let types = credential_type
+        .ok_or(inconsistent_jwt_claims("missing type-claim"))?;
+    let types = vc_type_entry
         .as_array()
-        .ok_or(inconsistent_jwt_claims("wrong types in JWT vc"))?;
-    if !types.contains(&Value::String(spec.credential_type.clone())) {
-        return Err(inconsistent_jwt_claims("wrong vc type"));
+        .ok_or(inconsistent_jwt_claims("malformed types-claim"))?;
+    if !types.contains(&Value::String(credential_type.clone())) {
+        return Err(inconsistent_jwt_claims(
+            "missing credential_type in type-claim",
+        ));
     };
+
+    // Check that credentialSubject-claim contains spec.credential_type entry with matching arguments.
     let credential_subject = vc_claims
         .get("credentialSubject")
+        .ok_or(inconsistent_jwt_claims("missing credentialSubject-claim"))?;
+    let subject = Subject::from_json_value(credential_subject.clone())
+        .map_err(|_| inconsistent_jwt_claims("malformed credentialSubject-claim"))?;
+    let verified_claim_arguments = subject
+        .properties
+        .get(credential_type)
+        .ok_or(inconsistent_jwt_claims("missing credential_type claim"))?
+        .as_object()
         .ok_or(inconsistent_jwt_claims(
-            "missing credentialSubject in JWT vc",
+            "malformed credential_type arguments",
         ))?;
-    let subject = Subject::from_json_value(credential_subject.clone()).map_err(|_| {
-        inconsistent_jwt_claims("missing credentialSubject in VerifiedAdult JWT vc")
-    })?;
-    let credential_type = &spec.credential_type;
-    let verified_claim_arguments = {
-        if let Some(claim) = subject.properties.get(credential_type) {
-            if let Some(claim_arguments) = claim.as_object() {
-                claim_arguments
-            } else {
-                return Err(inconsistent_jwt_claims(
-                    "wrong VerifiedAdult-claim in VerifiedAdult JWT vc",
-                ));
-            }
-        } else {
-            return Err(inconsistent_jwt_claims(
-                "missing VerifiedAdult-claim in VerifiedAdult JWT vc",
-            ));
-        }
+    let spec_arguments_count = if let Some(spec_arguments) = spec.arguments.as_ref() {
+        spec_arguments.len()
+    } else {
```

Comment:
This could maybe be solved a bit more concisely using `Option::map_or`?

Reply:
Indeed, done.